File: api/garmin/sync.js
// Vercel Serverless Function for Garmin Connect
export default async function handler(req, res) {
    res.setHeader('Access-Control-Allow-Origin', '*');
    if (req.method === 'OPTIONS') return res.status(200).end();
    if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' });

    const { email, password } = req.body;
    if (!email || !password) return res.status(400).json({ error: 'Credentials required' });

    try {
        const sso = await fetch('https://sso.garmin.com/sso/signin?service=https://connect.garmin.com/modern', {
            headers: { 'User-Agent': 'Mozilla/5.0' }
        });
        const cookies = sso.headers.get('set-cookie') || '';
        const html = await sso.text();
        const csrf = (html.match(/name="_csrf" value="([^"]+)"/) || [])[1] || '';

        const login = await fetch('https://sso.garmin.com/sso/signin?service=https://connect.garmin.com/modern', {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded', 'Cookie': cookies },
            body: new URLSearchParams({ username: email, password, embed: 'false', _csrf: csrf }),
            redirect: 'manual',
        });

        const ticket = ((await login.text()).match(/ticket=([^"&]+)/) || [])[1];
        if (!ticket) return res.status(401).json({ error: 'Login failed' });

        const session = await fetch('https://connect.garmin.com/modern/?ticket=' + ticket, {
            headers: { 'Cookie': cookies + '; ' + (login.headers.get('set-cookie') || '') },
            redirect: 'manual',
        });
        const allCookies = cookies + '; ' + (session.headers.get('set-cookie') || '');

        const today = new Date().toISOString().split('T')[0];
        const get = async (url) => {
            try { const r = await fetch(url, { headers: { 'Cookie': allCookies, 'NK': 'NT' } }); return r.ok ? r.json() : null; }
            catch { return null; }
        };

        const [sleep, hrv, sum] = await Promise.all([
            get('https://connect.garmin.com/modern/proxy/wellness-service/wellness/dailySleepData/' + today),
            get('https://connect.garmin.com/modern/proxy/hrv-service/hrv/' + today),
            get('https://connect.garmin.com/modern/proxy/usersummary-service/usersummary/daily/' + today),
        ]);

        return res.status(200).json({ success: true, data: {
            sleepHours: sleep?.dailySleepDTO?.sleepTimeSeconds ? Math.round(sleep.dailySleepDTO.sleepTimeSeconds/360)/10 : null,
            hrv: hrv?.hrvSummary?.lastNightAvg, hrvAvg7d: hrv?.hrvSummary?.weeklyAvg,
            restingHR: sum?.restingHeartRate, bodyBattery: sum?.bodyBatteryChargedValue,
            steps: sum?.totalSteps, lastSync: new Date().toISOString()
        }});
    } catch (e) { return res.status(500).json({ error: e.message }); }
}

File: api/garmin/sync.py
"""
Garmin Connect API - Python Serverless Function
Verwendet garminconnect library für zuverlässigen Sync
"""

from http.server import BaseHTTPRequestHandler
import json
import os
from datetime import datetime, timedelta

# Vercel KV für Token-Storage (oder Fallback auf File)
try:
    from garminconnect import Garmin, GarminConnectAuthenticationError
    GARMIN_AVAILABLE = True
except ImportError:
    GARMIN_AVAILABLE = False


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        """Handle CORS preflight"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_GET(self):
        """Health check endpoint"""
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()

        response = {
            'status': 'ok',
            'garmin_library': GARMIN_AVAILABLE,
            'timestamp': datetime.now().isoformat()
        }
        self.wfile.write(json.dumps(response).encode())

    def do_POST(self):
        """
        Sync Garmin data
        Body: { "email": "...", "password": "..." }
        Returns: { "success": true, "data": { sleep, hrv, bodyBattery, ... } }
        """
        self.send_header('Access-Control-Allow-Origin', '*')

        if not GARMIN_AVAILABLE:
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                'error': 'garminconnect library not available'
            }).encode())
            return

        # Parse request body
        content_length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(content_length)

        try:
            data = json.loads(body)
            email = data.get('email')
            password = data.get('password')

            if not email or not password:
                self.send_response(400)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({
                    'error': 'Email and password required'
                }).encode())
                return

            # Initialize Garmin client
            garmin = Garmin(email, password)

            try:
                garmin.login()
            except GarminConnectAuthenticationError as e:
                self.send_response(401)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({
                    'error': f'Authentication failed: {str(e)}'
                }).encode())
                return

            # Fetch today's data
            today = datetime.now().strftime('%Y-%m-%d')
            yesterday = (datetime.now() - timedelta(days=1)).strftime('%Y-%m-%d')

            result_data = {
                'lastSync': datetime.now().isoformat(),
                'sleepHours': None,
                'sleepScore': None,
                'bodyBattery': None,
                'hrv': None,
                'hrvAvg7d': None,
                'restingHR': None,
                'steps': None,
                'stressLevel': None
            }

            # Get sleep data
            try:
                sleep_data = garmin.get_sleep_data(today)
                if sleep_data:
                    sleep_seconds = sleep_data.get('dailySleepDTO', {}).get('sleepTimeSeconds', 0)
                    result_data['sleepHours'] = round(sleep_seconds / 3600, 1) if sleep_seconds else None
                    result_data['sleepScore'] = sleep_data.get('dailySleepDTO', {}).get('sleepScores', {}).get('overall', {}).get('value')
            except Exception as e:
                print(f"Sleep data error: {e}")

            # Get HRV data
            try:
                hrv_data = garmin.get_hrv_data(today)
                if hrv_data:
                    result_data['hrv'] = hrv_data.get('hrvSummary', {}).get('lastNightAvg')
                    result_data['hrvAvg7d'] = hrv_data.get('hrvSummary', {}).get('weeklyAvg')
            except Exception as e:
                print(f"HRV data error: {e}")

            # Get body battery
            try:
                bb_data = garmin.get_body_battery(today)
                if bb_data and len(bb_data) > 0:
                    latest = bb_data[-1] if isinstance(bb_data, list) else bb_data
                    result_data['bodyBattery'] = latest.get('bodyBatteryLevel') or latest.get('charged')
            except Exception as e:
                print(f"Body battery error: {e}")

            # Get resting heart rate
            try:
                hr_data = garmin.get_rhr_day(today)
                if hr_data:
                    result_data['restingHR'] = hr_data.get('restingHeartRate')
            except Exception as e:
                print(f"RHR data error: {e}")

            # Get steps from yesterday
            try:
                steps_data = garmin.get_steps_data(yesterday)
                if steps_data:
                    result_data['steps'] = steps_data.get('totalSteps')
            except Exception as e:
                print(f"Steps data error: {e}")

            # Get stress level
            try:
                stress_data = garmin.get_stress_data(today)
                if stress_data:
                    result_data['stressLevel'] = stress_data.get('overallStressLevel')
            except Exception as e:
                print(f"Stress data error: {e}")

            # Success response
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                'success': True,
                'data': result_data
            }).encode())

        except json.JSONDecodeError:
            self.send_response(400)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                'error': 'Invalid JSON'
            }).encode())
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                'error': str(e)
            }).encode())
